pkg/network/ebpf/c/prebuilt/tracer.c

```diff
@@ -363,7 +370,36 @@ static __always_inline bool check_family(struct sock* sk, u16 expected_family) {
     return family == expected_family;
 }
 
-static __always_inline int read_conn_tuple(conn_tuple_t* t, struct sock* skp, u64 pid_tgid, metadata_mask_t type) {
+static __always_inline int get_proto(conn_tuple_t * t) {
+    return (t->metadata & CONN_TYPE_TCP) ? CONN_TYPE_TCP : CONN_TYPE_UDP;
+}
+
+static __always_inline void sockaddr_to_addr(struct sockaddr * sa, u64 * addr_h, u64 * addr_l, u16 * port) {
+    if (!sa) return;
+
+    u16 family;
+    bpf_probe_read(&family, sizeof(family), &sa->sa_family);
+
+    struct sockaddr_in * sin;
+    struct sockaddr_in6 * sin6;
+    switch (family) {
+    case AF_INET:
+        bpf_probe_read(&sin, sizeof(sin), &sa);
+        if (addr_l) bpf_probe_read(addr_l, sizeof(u32), &(sin->sin_addr.s_addr));
+        if (port) bpf_probe_read(port, sizeof(u16), &sin->sin_port);
+        break;
+    case AF_INET6:
+        bpf_probe_read(&sin6, sizeof(sin6), &sa);
+        if (addr_l && addr_h) {
+            bpf_probe_read(addr_h, sizeof(u64), &(sin6->sin6_addr.s6_addr));
+            bpf_probe_read(addr_l, sizeof(u64), &(sin6->sin6_addr.s6_addr[8]));
+        }
+        if (port) bpf_probe_read(port, sizeof(u16), &sin6->sin6_port);
+        break;
+    }
+}
+
+static __always_inline int read_conn_tuple(conn_tuple_t* t, struct sockaddr * src_addr, struct sockaddr * dest_addr, struct sock* skp, u64 pid_tgid, metadata_mask_t type) {
```

Comment:
Not sure if splitting this would be better. Usually both src and dest are not available in `sockaddr` form. This function is for using `src_addr` and `dest_addr` first and then fall back to `sock` if the addresses are still not filled in.

Reply:
yeah, I gotta agree with @leeavital about this. this function is inlined in pretty much _all_ kprobes, so that's something to keep in mind.

Reply:
Ok, I took that code out from read_conn_tuple.